OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+  DataTable_<ET> rbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  /// Bind two DataTables by col.
+  template<typename ET>
+  DataTable_<ET> cbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  // Exceptions.
+  class ColumnDoesNotExist;
+  class ColumnHasLabel;
+  class ColumnHasNoLabel;
+  class ZeroElements;
+  class InvalidEntry;
+  class MetaDataKeyExists;
+  class MetaDataKeyDoesNotExist;
+  class MetaDataTypeMismatch;
+} // namespace OpenSim
+
+
+/**----------------------------------------------------------------------------*
+* AbstractDataTable is the base-class of all DataTable_ allowing storage of    *
+* DataTable_ of different types to be stored in containers. See Datatable_ for *
+* details on the interface.                                                    *
+*-----------------------------------------------------------------------------*/
+class OpenSim::AbstractDataTable {
+public:
+  AbstractDataTable() = default;
+  AbstractDataTable(const AbstractDataTable&) = delete;
+  AbstractDataTable(AbstractDataTable&&) = delete;
+  virtual std::unique_ptr<AbstractDataTable> clone() const = 0;
+  virtual ~AbstractDataTable() {}
+}; // AbstractDataTable
+
+class OpenSim::ColumnDoesNotExist : public std::runtime_error {
+public:
+  ColumnDoesNotExist(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ColumnHasLabel : public std::runtime_error {
+public:
+  ColumnHasLabel(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ColumnHasNoLabel : public std::runtime_error {
+public:
+  ColumnHasNoLabel(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::ZeroElements : public std::runtime_error {
+public:
+  ZeroElements(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::InvalidEntry : public std::runtime_error {
+public:
+  InvalidEntry(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataKeyExists : public std::runtime_error {
+public:
+  MetaDataKeyExists(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataKeyDoesNotExist : public std::runtime_error {
+public:
+  MetaDataKeyDoesNotExist(const std::string& expl) : runtime_error(expl) {}
+};
+
+class OpenSim::MetaDataTypeMismatch : public std::runtime_error {
+public:
+  MetaDataTypeMismatch(const std::string& expl) : runtime_error(expl) {}
+};
+
+
+/**----------------------------------------------------------------------------*
+* \brief DataTable_ is a in-memory storage container for data(in the form of a *
+* matrix with column names) with support for holding metatdata.                *
+*                                                                              *
+* <ul> <li>Underlying matrix will have entries of configurable type            *
+* ET(template param).</li>  <li>Random-access(constant-time) to specific       *
```

Comment:
There is a much better way to do lists in Markdown-enabled doxygen: just being each line with a dash like this:
```
- Underlying matrix will have entries of configurable type ET(template param).
- Random-access(constant-time) to specific entries
```
Numbered lists work too. The greatly reduces the amount of metadata (i.e., clutter) and makes the formatting the same for programmers reading the code as for users reading the doxygen.